src/app/controllers/__tests__/frontend.server.controller.spec.ts

```diff
@@ -0,0 +1,80 @@
+import { StatusCodes } from 'http-status-codes'
+
+import expressHandler from 'tests/unit/backend/helpers/jest-express'
+
+import frontendServerController from '../frontend.server.controller'
+
+describe('frontend.server.controller', () => {
+  afterEach(() => jest.clearAllMocks())
+  const mockRes = expressHandler.mockResponse()
+  const mockReq = {
+    app: {
+      locals: {
+        GATrackingID: 'abc',
+        appName: 'xyz',
+        environment: 'efg',
+      },
+    },
+    query: {
+      redirectPath: 'formId?fieldId1=abc&fieldId2=xyz',
+    },
+  }
+  describe('datalayer', () => {
+    it('should return the correct value', () => {
+      frontendServerController.datalayer(mockReq, mockRes)
+      expect(mockRes.send).toHaveBeenCalledWith(
+        expect.stringContaining("'app_name': 'xyz'"),
+      )
+      expect(mockRes.send).toHaveBeenCalledWith(
+        expect.stringContaining("'config', 'abc'"),
+      )
+    })
+
+    it('should call type and status correctly', () => {
+      frontendServerController.datalayer(mockReq, mockRes)
+      expect(mockRes.type).toHaveBeenCalledWith('text/javascript')
+      expect(mockRes.status).toHaveBeenCalledWith(StatusCodes.OK)
```

Comment:
a good way to think about tests is that each `it` should check that the code has the right outputs for a _fixed set of inputs_, instead of fixing the inputs and having multiple `it` blocks for different parts of the output. this makes it much easier to debug and organise tests.

for example, if we were to add error-handling code to `frontend.server.controller` (which we really should for `ejs.render`), how would you add more tests here? if we were to continue following the current pattern, we would add two tests, each with the same input and testing for different parts of the output. but then we'd have 4 tests, each with a different set of inputs and outputs, and you can imagine how disorganised the tests might get as the module gets more complex.

instead, I think it would be better to merge all these `expect` statements into a single test, and call it something like `should return the correct response when inputs are valid`. then if the test ever fails, it's immediately clear what inputs have caused the test to fail, and all the expected outputs can be seen in one place.